cryostatio/cryostat
# feat(discoveryplugin): look up plugin callback credentials in database

# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
_______________________________________________

Fixes: #1376

## Description of the change:
~~If a discovery plugin (ex. the Agent) has registered using a callback URL that includes userinfo (ex. `http://user:pass@example.com`),
this same userinfo should be included by Cryostat both in the request URL as well as copied to `Authorization` header when Cryostat communicates with that plugin, for example when pinging the plugin to ensure it is still live and reachable.~~

When a discovery plugin wants to register with Cryostat, it will first define a stored credential for itself. Then it will register itself using a special value for the userinfo of the plugin callback URI which references that stored credential. Cryostat looks up that credential and includes it for authentication on the plugin callback requests it makes in the future, both to validate the plugin during registration time as well as periodically later for heartbeat checks.

## Motivation for the change:
Plugins should require an authentication challenge on requests since these requests can trigger the plugin to refresh registration with the Cryostat server. Without an authentication challenge, any client may ping the plugin instance and cause it to resend registration to Cryostat. This causes undue increased load on the Cryostat server for no productive purpose.

## How to manually test:
1. Check out this PR image
2. Check out [associated Agent PR](https://github.com/cryostatio/cryostat-agent/pull/64), build agent into local m2 repo, and rebuild `quarkus-test` container image
3. Run `smoketest.sh` using this PR image and the updated `quarkus-test` with updated `-agent`
4. Verify from logs and target discovery queries that the agent has registered and published itself with generated basic credentials, and that the agent requires these credentials to be passed with requests.
